k8s/module-artifact-gen/k8s-artifact-gen

```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+set -e
+
+show_help() {
+  cat << EOF
+
+Simple tool to generate Mender Artifact suitable for k8s Update Module
+
+Usage: $0 [options] MANIFESTO [MANIFESTOS...] [-- [options-for-mender-artifact] ]
+
+    Options: [ -n|artifact-name -t|--device-type -o|--output_path -h|--help ]
+
+        --artifact-name     - Artifact name
+        --device-type       - Target device type identification (can be given more than once)
+        --output-path       - Path to output file. Default: k8s-artifact.mender
+        --help              - Show help and exit
+        MANIFESTO           - K8s manifesto(s) to add to the Artifact
+
+Anything after a '--' gets passed directly to the mender-artifact tool.
+
+EOF
+}
+
+show_help_and_exit_error() {
+  show_help
+  exit 1
+}
+
+check_dependency() {
+  hash "$1" || exit 1
+}
+
+check_dependency mender-artifact
+check_dependency kubectl
+
+device_types=""
+artifact_name=""
+output_path="k8s-artifact.mender"
+MANIFESTOS=""
+passthrough=0
+passthrough_args=""
+
+while (( "$#" )); do
+  if test $passthrough -eq 1
+  then
+    passthrough_args="$passthrough_args $1"
+    shift
+    continue
+  fi
+  case "$1" in
+    --device-type | -t)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      device_types="$device_types $1 $2"
+      shift 2
+      ;;
+    --artifact-name | -n)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      artifact_name=$2
+      shift 2
+      ;;
+    --output-path | -o)
+      if [ -z "$2" ]; then
+        show_help_and_exit_error
+      fi
+      output_path=$2
+      shift 2
+      ;;
+    -h | --help)
+      show_help
+      exit 0
+      ;;
+    --)
+      passthrough=1
+      shift
+      ;;
+    -*)
+      echo "Error: unsupported option $1"
+      show_help_and_exit_error
+      ;;
+    *)
+      MANIFESTOS="$MANIFESTOS $1"
+      shift
+      ;;
+  esac
+done
+
+if [ -z "${artifact_name}" ]; then
+  echo "Artifact name not specified. Aborting."
+  show_help_and_exit_error
+fi
+
+if [ -z "${device_types}" ]; then
+  echo "Device type not specified. Aborting."
+  show_help_and_exit_error
+fi
+
+if [ -z "${MANIFESTOS}" ]; then
+  echo "At least one k8s manifesto must be specified. Aborting."
+  show_help_and_exit_error
+fi
+
+manifestos=""
+for manifesto in $MANIFESTOS; do
+  if ! kubectl apply --dry-run -f $manifesto >/dev/null 2>&1; then
+    echo "The ${manifesto} k8s manifesto is not valid. Aborting."
+    show_help_and_exit_error
+  fi
+  manifestos="$manifestos -f ${manifesto}"
+done
```

Comment:
Admittedly I don't know so much about k8s, so I could be completely wrong, but here's the question: In the `docker` Update Module, we extract the SHA of the container to make sure that the update is completely stable, and won't change even if the tag moves. For example from `ubuntu:18.04` we would extract its SHA and use that instead of the name. Is that something that could/should be done for the manifesto?

Reply:
@kacf this doesn't apply to Kubernetes: what you are applying is not the execution of a docker image, but the manifesto which is stored in etcd (or other storage, for different implementation than the stock Kubernetes). It can include the docker images (if it's a pod, deployment, deamonset, whatever) but it can also describe a service, a persistent volume, an ingress, a secret, any other kind of kubernetes resource you want to set.

Kubernetes is a declarative tool: you describe the infrastructure using manifestos, the engine is responsible for reaching a stable state which fulfils the requirements you expressed with your manifesto (e.g., 2 replicas of a given set of containers, which we call pods, with a balancer in front of it).

To wrap it up: no, I don't think we have to mangle the manifesto in any way and deal with SHA of docker images. The images themselves are expected to be pulled from registries, either public or private (k8s manifestos can describe credentials to access protected registries if needed). The delivery of a binary docker image is a different use case, and it requires a specific property in the manifesto avoiding pulling from a remote registry. It makes sense to have an update module to delivery binary docker images as part of the update process, but should be independent from the k8s one.

I hope this answers your question. :)

Reply:
Perhaps you misunderstood, we are not delivering binary docker images with the `docker` update module either. It is pulling from registries, just like k8s. The difference is that it's pulling a specific SHA, instead of an image tag. If I understood correctly, this is the problem that Alex was having before we started tagging specifically for Hosted Mender, that containers like `mendersoftware:deployments:1.7.0` would change behind his back, and two replicas may actually be different, depending on when they were created. This is also the reason why we are now changing to Docker tags that reflect the Mender Product version, instead of the component versions.

So the mangling would only be a textual change in the manifesto, to point to a SHA instead of a tag, and the image would be pulled from the registry.

But if you still think it's inappropriate, I'm fine with that, I just wanted to bring it up!

Reply:
I know the docker update module is not delivering the binary image, I mentioned it as a different use case which can make sense (maybe using a specific update module? not sure) which can be complementary to k8s and docker.

There is an important difference between the docker and the k8s update modules, though: in docker you are running a container, and the module makes sure you are going to run exactly the container you intended to run using the SHA to identify it. The k8s update module applies a kubernetes manifesto, which you expect to be applied as it is. If you want to run a specific docker image, you fix the tag in the yaml and you define the correct pullPolicy. I don't know if it makes sense, but it is more like installing a config file: I do not expect the Mender client to mangle my configuration file when installing it into the device.

Reply:
Alright, makes sense, then I won't argue any further! :-)